### ISSUE TITLE
[Feature] Bidirectionnal

### ISSUE THREAD

**Issue by thegodone** [ml-explore/mlx#983]
Can we add the bidirectionnal layer for RNN like LSTM or GRU ?

**Comment by awni** [comment]
I don't think we will add yet, as for now it's easy enough to do:

```python
h_forward = RNN(x)
h_backward = BackRNN(x[:, ::-1, :])
h = h_forward + h_backward # or stack them.. depending what you want.
```

And different architectures have different preferences for how / where they stitch the forward and backward states together.

**Comment by thegodone** [comment]
and for TimeDistributed ?

**Comment by thegodone** [comment]
What's about self attention like class under, can I use MultiHeadAttention ?
## Custom attention layer
# modified from https://github.com/sujitpal/eeap-examples
class AttentionM(Layer):
    """
    Keras layer to compute an attention vector on an incoming matrix.
    # Input
        enc - 3D Tensor of shape (BATCH_SIZE, MAX_TIMESTEPS, EMBED_SIZE)
    # Output
        2D Tensor of shape (BATCH_SIZE, EMBED_SIZE)
    # Usage
        enc = LSTM(EMBED_SIZE, return_sequences=True)(...)
        att = AttentionM()(enc)
    """
    def __init__(self, axis=-1, return_probabilities = False, **kwargs):
        self.return_probabilities = return_probabilities
        self.axis = axis
        super(AttentionM, self).__init__(**kwargs)

    
    def build(self, input_shape):
        # W: (EMBED_SIZE, 1)
        # b: (MAX_TIMESTEPS,)
        self.W = self.add_weight(name="W_{:s}".format(self.name),
                                 shape=(input_shape[-1], 1),
                                 initializer="normal")
        self.b = self.add_weight(name="b_{:s}".format(self.name),
                                 shape=(input_shape[1], 1),
                                 initializer="zeros")
        super(AttentionM, self).build(input_shape)


    def call(self, x, mask=None):
        # input: (BATCH_SIZE, MAX_TIMESTEPS, EMBED_SIZE)
        # et: (BATCH_SIZE, MAX_TIMESTEPS)
        et = K.squeeze(K.tanh(K.dot(x, self.W) + self.b), axis=self.axis)
        # at: (BATCH_SIZE, MAX_TIMESTEPS)
        at = K.softmax(et)
        if mask is not None:
            at *= K.cast(mask, K.floatx())
        # atx: (BATCH_SIZE, MAX_TIMESTEPS, 1)
        atx = K.expand_dims(at, axis=-1)
        # ot: (BATCH_SIZE, MAX_TIMESTEPS, EMBED_SIZE)
        ot = x * atx
        # output: (BATCH_SIZE, EMBED_SIZE)
        if self.return_probabilities:
            return atx # for visualization of the attention weights
        else:
            return K.sum(ot, axis=1) # for prediction

    def compute_mask(self, input, input_mask=None):
        # do not pass the mask to the next layers
        return None
    
    
    def compute_output_shape(self, input_shape):
        return (input_shape[0], input_shape[-1])


    def get_config(self):
        return super(AttentionM, self).get_config()

**Comment by awni** [comment]
We have it: https://ml-explore.github.io/mlx/build/html/python/nn/_autosummary/mlx.nn.MultiHeadAttention.html#mlx.nn.MultiHeadAttention
